From 8e28fd505997ff634fa654203c6aeeefcbfd00eb Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 15:22:18 -0400
Subject: [infiniband] ocrdma: add default gid at index 0

Message-id: <4bc74b8b51ef5da189fbc0e075b22dcf988b6a0a.1411571481.git.dledford@redhat.com>
Patchwork-id: 96647
O-Subject: [Patch RHEL7 43/48] RDMA/ocrdma: add default gid at index 0
Bugzilla: 1081067 1088577
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Devesh Sharma <devesh.sharma@emulex.com>

This patch adds the code to initialize default gid
at index 0. Rest of the ip4 and ip6 gid addition
remains as it is.

Signed-off-by: Devesh Sharma <devesh.sharma@emulex.com>
(cherry picked from commit 2f7a3d2abfd786bbbb32e664b6ff58daf1d338c2)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/ocrdma/ocrdma_main.c b/drivers/infiniband/hw/ocrdma/ocrdma_main.c
index 256a06b..cb7b52a 100644
--- a/drivers/infiniband/hw/ocrdma/ocrdma_main.c
+++ b/drivers/infiniband/hw/ocrdma/ocrdma_main.c
@@ -388,6 +388,15 @@ static void ocrdma_remove_sysfiles(struct ocrdma_dev *dev)
   device_remove_file(&dev->ibdev.dev, ocrdma_attributes[i]);
 }
 
+static void ocrdma_add_default_sgid(struct ocrdma_dev *dev)
+{
+ /* GID Index 0 - Invariant manufacturer-assigned EUI-64 */
+ union ib_gid *sgid = &dev->sgid_tbl[0];
+
+ sgid->global.subnet_prefix = cpu_to_be64(0xfe80000000000000LL);
+ ocrdma_get_guid(dev, &sgid->raw[8]);
+}
+
 static void ocrdma_init_ipv4_gids(struct ocrdma_dev *dev,
       struct net_device *net)
 {
@@ -434,6 +443,7 @@ static void ocrdma_init_gid_table(struct ocrdma_dev *dev)
     rdma_vlan_dev_real_dev(net_dev) : net_dev;
 
   if (real_dev == dev->nic_info.netdev) {
+   ocrdma_add_default_sgid(dev);
    ocrdma_init_ipv4_gids(dev, net_dev);
    ocrdma_init_ipv6_gids(dev, net_dev);
   }
-- 
1.7.1